nifi-nar-bundles/nifi-sql-reporting-bundle/nifi-sql-reporting-tasks/src/main/java/org/apache/nifi/reporting/sql/MetricsEventReportingTask.java

```diff
@@ -77,17 +83,37 @@ public void setup(final ConfigurationContext context) throws IOException {
 
     @Override
     public void onTrigger(ReportingContext context) {
+        String sql = context.getProperty(QueryMetricsUtil.QUERY).evaluateAttributeExpressions().getValue();
```

Comment:
The duplication of the sql string start time and end time evaluation logic could be prevented.
I suggest a trait-like interface to achieve maximum potential like this:
```java
public class MetricsEventReportingTask extends AbstractReportingTask implements QueryTimeAware {

    private List<PropertyDescriptor> properties;
    private MetricsQueryService metricsQueryService;
    private volatile RulesEngineService rulesEngineService;
    private volatile PropertyContextActionHandler actionHandler;

    @Override
    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
        return properties;
    }

    @Override
    protected void init(final ReportingInitializationContext config) {
        final List<PropertyDescriptor> properties = new ArrayList<>();
        properties.add(QueryMetricsUtil.QUERY);
        properties.add(QueryMetricsUtil.RULES_ENGINE);
        properties.add(QueryMetricsUtil.ACTION_HANDLER);
        properties.add(VARIABLE_REGISTRY_ONLY_DEFAULT_PRECISION);
        properties.add(VARIABLE_REGISTRY_ONLY_DEFAULT_SCALE);
        this.properties = Collections.unmodifiableList(properties);
    }

    @OnScheduled
    public void setup(final ConfigurationContext context) {
        actionHandler = context.getProperty(QueryMetricsUtil.ACTION_HANDLER).asControllerService(PropertyContextActionHandler.class);
        rulesEngineService = context.getProperty(QueryMetricsUtil.RULES_ENGINE).asControllerService(RulesEngineService.class);
        final Integer defaultPrecision = context.getProperty(VARIABLE_REGISTRY_ONLY_DEFAULT_PRECISION).evaluateAttributeExpressions().asInteger();
        final Integer defaultScale = context.getProperty(VARIABLE_REGISTRY_ONLY_DEFAULT_SCALE).evaluateAttributeExpressions().asInteger();
        metricsQueryService = new MetricsSqlQueryService(getLogger(), defaultPrecision, defaultScale);
    }

    @Override
    public void onTrigger(ReportingContext context) {
        String sql = context.getProperty(QueryMetricsUtil.QUERY).evaluateAttributeExpressions().getValue();
        try {
            sql = processStartAndEndTimes(context, sql, BULLETIN_START_TIME, BULLETIN_END_TIME);
            sql = processStartAndEndTimes(context, sql, PROVENANCE_START_TIME, PROVENANCE_END_TIME);

            fireRules(context, actionHandler, rulesEngineService, sql);
        } catch (Exception e) {
            getLogger().error("Error opening loading rules: {}", new Object[]{e.getMessage()}, e);
        }
    }

    private void fireRules(ReportingContext context, PropertyContextActionHandler actionHandler, RulesEngineService engine, String query) throws Exception {
        getLogger().debug("Executing query: {}", query);
        QueryResult queryResult = metricsQueryService.query(context, query);
        ResultSetRecordSet recordSet = metricsQueryService.getResultSetRecordSet(queryResult);
        Record record;
        try {
            while ((record = recordSet.next()) != null) {
                final Map<String, Object> facts = new HashMap<>();
                for (String fieldName : record.getRawFieldNames()) {
                    facts.put(fieldName, record.getValue(fieldName));
                }
                List<Action> actions = engine.fireRules(facts);
                if (actions == null || actions.isEmpty()) {
                    getLogger().debug("No actions required for provided facts.");
                } else {
                    actions.forEach(action -> actionHandler.execute(context, action, facts));
                }
            }
        } finally {
            metricsQueryService.closeQuietly(recordSet);
        }
    }
}

public class QueryNiFiReportingTask extends AbstractReportingTask implements QueryTimeAware {

    private List<PropertyDescriptor> properties;

    private volatile RecordSinkService recordSinkService;

    private MetricsQueryService metricsQueryService;

    @Override
    protected void init(final ReportingInitializationContext config) {
        final List<PropertyDescriptor> properties = new ArrayList<>();
        properties.add(QueryMetricsUtil.QUERY);
        properties.add(QueryMetricsUtil.RECORD_SINK);
        properties.add(QueryMetricsUtil.INCLUDE_ZERO_RECORD_RESULTS);
        properties.add(VARIABLE_REGISTRY_ONLY_DEFAULT_PRECISION);
        properties.add(VARIABLE_REGISTRY_ONLY_DEFAULT_SCALE);
        this.properties = Collections.unmodifiableList(properties);
    }

    @Override
    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
        return properties;
    }

    @OnScheduled
    public void setup(final ConfigurationContext context) {
        recordSinkService = context.getProperty(QueryMetricsUtil.RECORD_SINK).asControllerService(RecordSinkService.class);
        recordSinkService.reset();
        final Integer defaultPrecision = context.getProperty(VARIABLE_REGISTRY_ONLY_DEFAULT_PRECISION).evaluateAttributeExpressions().asInteger();
        final Integer defaultScale = context.getProperty(VARIABLE_REGISTRY_ONLY_DEFAULT_SCALE).evaluateAttributeExpressions().asInteger();
        metricsQueryService = new MetricsSqlQueryService(getLogger(), defaultPrecision, defaultScale);
    }

    @Override
    public void onTrigger(ReportingContext context) {
        final StopWatch stopWatch = new StopWatch(true);
        String sql = context.getProperty(QueryMetricsUtil.QUERY).getValue();
        try {
            sql = processStartAndEndTimes(context, sql, BULLETIN_START_TIME, BULLETIN_END_TIME);
            sql = processStartAndEndTimes(context, sql, PROVENANCE_START_TIME, PROVENANCE_END_TIME);

            getLogger().debug("Executing query: {}", sql);
            final QueryResult queryResult = metricsQueryService.query(context, sql);
            final ResultSetRecordSet recordSet;

            try {
                recordSet = metricsQueryService.getResultSetRecordSet(queryResult);
            } catch (final Exception e) {
                getLogger().error("Error creating record set from query results due to {}", new Object[]{e.getMessage()}, e);
                return;
            }

            try {
                final Map<String, String> attributes = new HashMap<>();
                final String transactionId = UUID.randomUUID().toString();
                attributes.put("reporting.task.transaction.id", transactionId);
                attributes.put("reporting.task.name", getName());
                attributes.put("reporting.task.uuid", getIdentifier());
                attributes.put("reporting.task.type", this.getClass().getSimpleName());
                recordSinkService.sendData(recordSet, attributes, context.getProperty(QueryMetricsUtil.INCLUDE_ZERO_RECORD_RESULTS).asBoolean());
            } catch (Exception e) {
                getLogger().error("Error during transmission of query results due to {}", new Object[]{e.getMessage()}, e);
                return;
            } finally {
                metricsQueryService.closeQuietly(queryResult);
            }
            final long elapsedMillis = stopWatch.getElapsed(TimeUnit.MILLISECONDS);
            getLogger().debug("Successfully queried and sent in {} millis", elapsedMillis);
        } catch (Exception e) {
            getLogger().error("Error processing the query due to {}", new Object[]{e.getMessage()}, e);
        }
    }
}

public interface QueryTimeAware {
    default String processStartAndEndTimes(
        ReportingContext context,
        String sql,
        TrackedQueryTime queryStartTime,
        TrackedQueryTime queryEndTime
    ) throws IOException {
        StateManager stateManager = context.getStateManager();

        final Map<String, String> stateMap = new HashMap<>(stateManager.getState(Scope.LOCAL).toMap());

        if (sql.contains(queryStartTime.getSqlPlaceholder()) && sql.contains(queryEndTime.getSqlPlaceholder())) {
            final long startTime = stateMap.get(queryStartTime.name()) == null ? 0 : Long.parseLong(stateMap.get(queryStartTime.name()));
            final long currentTime = getCurrentTime();

            sql = sql.replace(queryStartTime.getSqlPlaceholder(), String.valueOf(startTime));
            sql = sql.replace(queryEndTime.getSqlPlaceholder(), String.valueOf(currentTime));

            stateMap.put(queryStartTime.name(), String.valueOf(currentTime));
            stateManager.setState(stateMap, Scope.LOCAL);
        }

        return sql;
    }

    default long getCurrentTime() {
        return Instant.now().toEpochMilli();
    }
}

```